TAO/orbsvcs/orbsvcs/Event/ECG_Mcast_EH.cpp

```diff
@@ -197,6 +200,7 @@ int
 TAO_ECG_Mcast_EH::delete_unwanted_subscriptions (
     Address_Set& multicast_addresses)
 {
+  ACE_GUARD_RETURN (TAO_SYNCH_MUTEX, ace_mon, this->lock_, -1);
```

Comment:
OK, so here's my additional thoughts:

The real problem is that removing members from the subscription array reorders the array. So it seems the delete_unwanted_members should be refactored so that it is a two-phase process: First, while under the lock, collect the entries from subscription array that need to be removed to a new stack-based list, reordering the subscriptions array at will. Then after releasing the lock on the subscription array, you can safely clean up the resources of the removed members.

While in handle_input, use the guard to lock the subscription array while searching, but when the correct receiver is found, use guard.release() to unlock the array again before calling the receiver's handle input.

Note that this still makes it possible for a subscription member to be closed down while its associated peer data is being received, but that's a separate issue. 

As for the pull request title, how about something like "Synchronize access to the ECH Mcast subscription list?"